src/biotite/sequence/profile.py

```diff
@@ -0,0 +1,211 @@
+
+        Based on the sequences of the alignment, the SequenceProfile parameters symbols
+        and gaps are calculated.
+        """
+        sequences = get_codes(alignment)
+        for i in range(1, len(sequences)):
+            if not set(alignment.sequences[i-1].get_alphabet().get_symbols()) == \
+                   set(alignment.sequences[i].get_alphabet().get_symbols()):
+                raise TypeError("Alignment contains sequences with different alphabets")
+        alphabet = alignment.sequences[0].get_alphabet().get_symbols()
+        symbols = np.zeros((len(sequences[0]), len(alphabet)), dtype=int)
+        gaps = np.zeros(len(sequences[0]), dtype=int)
+        sequences = np.transpose(sequences)
+        for i in range(len(sequences)):
+            row = np.where(sequences[i, ] == -1, len(alphabet), sequences[i, ])
+            if len(np.bincount(row)) < len(alphabet):
+                symbols[i, 0:len(np.bincount(row))] = np.bincount(row)
+            elif len(np.bincount(row)) > len(alphabet):
+                symbols[i, ] = np.bincount(row)[0:len(alphabet)]
+                gaps[i] = np.bincount(row)[-1]
+            else:
+                symbols[i, ] = np.bincount(row)
+        return SequenceProfile(symbols, gaps, alignment.sequences[0].get_alphabet())
+
+    def to_consensus(self, as_general=False):
+        """
+        Get the consensus sequence for this SequenceProfile object.
+
+        Parameters
+        ----------
+        as_general : bool
+            If true, returns consensus sequence as GeneralSequence object.
+            Otherwise, the consensus sequence object type is chosen based
+            on the alphabet of this SequenceProfile object.(Default: False).
+
+        Returns
+        -------
+        consensus: either NucleotideSequence, ProteinSequence or GeneralSequence
+            The calculated consensus sequence
+        """
+        # https://en.wikipedia.org/wiki/International_Union_of_Pure_and_Applied_Chemistry#Amino_acid_and_nucleotide_base_codes
+        if as_general:
+            return GeneralSequence(self.alphabet, self._custom_to_consensus())
+        elif set(self.alphabet.get_symbols()) == set(_NUC_DNA_ALPH.get_symbols()):
+            return NucleotideSequence(self._dna_to_consensus())
+        elif set(self.alphabet.get_symbols()) == set(_NUC_RNA_ALPH.get_symbols()):
+            return NucleotideSequence(self._rna_to_consensus())
+        elif set(self.alphabet.get_symbols()) == set(_PROT_ALPH.get_symbols()):
+            return ProteinSequence(self._prot_to_consensus())
+        return GeneralSequence(self.alphabet, self._custom_to_consensus())
+
+    def _codes_to_iupac(self, frequency, codes):
+        if np.sum(frequency) == 0:
+            return "-"
+        position_max = np.where(frequency == np.amax(frequency))[0].tolist()
+        key = "".join(["".join(self.alphabet.get_symbols()[i]) for i in position_max])
```

Comment:
Calling `self.alphabet.get_symbols()` for each column introduces some overhead, since the list of symbols is copied everytime. it is probably more efficient to use `Alphabet.decode()` instead or, even better, to directly use symbols codes and skip using characters entirely.